src/syncManager.ts

```diff
@@ -0,0 +1,116 @@
+    @inject(Services.CRYPTO_CONFIG) private readonly cryptoConfig: ICryptoConfig,
+    private readonly queueClient: QueueClient,
+    private readonly tilesManager: TilesManager,
+    private readonly cryptoManager: CryptoManager,
+    private readonly nifiClient: NifiClient
+  ) {
+    this.logger = logger;
+    this.syncAttempts = this.config.get<number>('syncAttempts');
+  }
+
+  public async runSync(): Promise<void> {
+    const task = await this.queueClient.queueHandler.waitForTask();
+
+    if (task) {
+      const params = task.parameters as IParameters;
+      const jobId = task.jobId;
+      const taskId = task.id;
+      const batch = params.batch;
+      const attempts = task.attempts;
+      const layerId = `${params.resourceId}-${params.resourceVersion}`;
+
+      if (attempts <= this.syncAttempts) {
+        try {
+          this.logger.debug(`tiles signature is set to: ${this.tilesConfig.sigIsNeeded.toString()}`);
+          const generator = tilesGenerator(batch);
+          let batchArray = [];
+
+          for (const tile of generator) {
+            const path = `${this.tilesConfig.path}/${layerId}/${tile.zoom}/${tile.x}/${tile.y}.${this.tilesConfig.format}`;
+
+            if (await isFileExists(path)) {
+              batchArray.push(this.signAndUpload(tile, path));
+            }
+            if (batchArray.length === this.tilesConfig.uploadBatchSize) {
+              await Promise.all(batchArray);
+              await this.tilesManager.updateTilesCount(layerId, batchArray.length);
+              batchArray = [];
+            }
+            // resolved left overs
+            await Promise.all(batchArray);
+          }
+
+          await this.tilesManager.updateTilesCount(layerId, batchArray.length);
+
+          await this.queueClient.queueHandler.ack(jobId, taskId);
+          await this.finishJob(jobId, layerId);
+        } catch (error) {
+          await this.queueClient.queueHandler.reject(jobId, taskId, true, (error as Error).message);
+        }
+      } else {
+        await this.queueClient.queueHandler.reject(jobId, taskId, false);
+        await this.finishJob(jobId, layerId, false);
+      }
+    }
+  }
+
+  private async signAndUpload(tile: ITile, path: string): Promise<void> {
+    if (this.tilesConfig.sigIsNeeded) {
+      await this.cryptoManager.generateSignedFile(this.cryptoConfig.pem, path);
+    }
+    await this.tilesManager.uploadTile(tile, path);
+  }
+
+  private async finishJob(jobId: string, layerId: string, isSuccess = true, errorReason: string | undefined = undefined): Promise<void> {
+    this.logger.info(`Update Job status to success=${String(isSuccess)} jobId=${jobId}`);
+    const joUpdatePayload: IUpdateJobRequestPayload = {
+      status: isSuccess ? TaskStatus.COMPLETED : TaskStatus.FAILED,
+      reason: errorReason,
+    };
+    await this.queueClient.queueHandler.jobManagerClient.updateJob(jobId, joUpdatePayload);
```

Comment:
i think the ni-fi should determine the job status - talk with me

Reply:
fixed